File: js/src/P2P2Core.ts
// Core P2P functionality that can be used in any environment
import { RoomConfig } from './types';
import { CloudflareDNSDiscovery } from './CloudflareDNSDiscovery';
import { WebRTCManager } from './WebRTCManager';
import { EnvironmentAdapter, createDefaultAdapter } from './adapters/EnvironmentAdapter';

export interface P2P2Options {
  adapter?: EnvironmentAdapter;
  discoveryInterval?: number;
  iceServers?: RTCIceServer[];
}

export class P2P2Core {
  protected dnsDiscovery: CloudflareDNSDiscovery;
  protected webRTCManager: WebRTCManager;
  protected adapter: EnvironmentAdapter;
  protected discoveryInterval: number;

  constructor(config: RoomConfig, options: P2P2Options = {}) {
    this.adapter = options.adapter || createDefaultAdapter();
    this.discoveryInterval = options.discoveryInterval || 5000;
    
    this.dnsDiscovery = new CloudflareDNSDiscovery(
      config.domain,
      config.zoneId,
      config.apiToken,
      this.adapter
    );
    
    this.webRTCManager = new WebRTCManager(config);
  }

  getPeerId(): string {
    return this.dnsDiscovery.getPeerId();
  }

  // Core DNS operations
  async announcePresence(roomId: string): Promise<void> {
    return this.dnsDiscovery.announcePresence(roomId);
  }

  async removePresence(roomId: string): Promise<void> {
    return this.dnsDiscovery.removePresence(roomId);
  }

  async discoverPeers(roomId: string): Promise<string[]> {
    return this.dnsDiscovery.discoverPeers(roomId);
  }

  async publishSignalingData(roomId: string, targetPeerId: string, data: string): Promise<void> {
    return this.dnsDiscovery.publishSignalingData(roomId, targetPeerId, data);
  }

  async getSignalingData(roomId: string, fromPeerId: string): Promise<string | null> {
    return this.dnsDiscovery.getSignalingData(roomId, fromPeerId);
  }

  // Core WebRTC operations
  async createPeerConnection(): Promise<RTCPeerConnection> {
    return this.webRTCManager.createPeerConnection();
  }

  createDataChannel(connection: RTCPeerConnection, label: string = 'data'): RTCDataChannel {
    return this.webRTCManager.createDataChannel(connection, label);
  }

  async createOffer(connection: RTCPeerConnection): Promise<RTCSessionDescriptionInit> {
    return this.webRTCManager.createOffer(connection);
  }

  async createAnswer(connection: RTCPeerConnection): Promise<RTCSessionDescriptionInit> {
    return this.webRTCManager.createAnswer(connection);
  }

  async setRemoteDescription(
    connection: RTCPeerConnection,
    description: RTCSessionDescriptionInit
  ): Promise<void> {
    return this.webRTCManager.setRemoteDescription(connection, description);
  }

  async addIceCandidate(
    connection: RTCPeerConnection,
    candidate: RTCIceCandidateInit
  ): Promise<void> {
    return this.webRTCManager.addIceCandidate(connection, candidate);
  }

  // Timer utilities using adapter
  setInterval(callback: () => void, ms: number): any {
    return this.adapter.setInterval(callback, ms);
  }

  clearInterval(handle: any): void {
    this.adapter.clearInterval(handle);
  }

  setTimeout(callback: () => void, ms: number): any {
    return this.adapter.setTimeout(callback, ms);
  }

  clearTimeout(handle: any): void {
    this.adapter.clearTimeout(handle);
  }

  async waitForIceGathering(connection: RTCPeerConnection): Promise<void> {
    return this.webRTCManager.waitForIceGathering(connection);
  }
}

File: Sources/P2P2/P2P2Core.swift
import Foundation
@preconcurrency import WebRTC

/// Core P2P functionality that can be used in any Swift environment
public final class P2P2Core {
    public let dnsDiscovery: CloudflareDNSDiscovery
    public let webRTCManager: WebRTCManager
    
    public init(config: RoomConfig) {
        self.webRTCManager = WebRTCManager(
            stunServers: config.stunServers,
            turnServers: config.turnServers
        )
        self.dnsDiscovery = CloudflareDNSDiscovery(
            domain: config.domain,
            zoneId: config.zoneId,
            apiToken: config.apiToken
        )
    }
    
    // MARK: - DNS Operations
    
    public func getPeerId() -> String {
        return dnsDiscovery.peerId
    }
    
    public func announcePresence(roomId: String) async throws {
        try await dnsDiscovery.announcePresence(roomId: roomId)
    }
    
    public func removePresence(roomId: String) async throws {
        try await dnsDiscovery.removePresence(roomId: roomId)
    }
    
    public func discoverPeers(roomId: String) async throws -> [String] {
        return try await dnsDiscovery.discoverPeers(roomId: roomId)
    }
    
    public func publishSignalingData(roomId: String, targetPeerId: String, data: String) async throws {
        try await dnsDiscovery.publishSignalingData(
            roomId: roomId,
            peerId: targetPeerId,
            data: data
        )
    }
    
    public func getSignalingData(roomId: String, fromPeerId: String) async throws -> String? {
        return try await dnsDiscovery.getSignalingData(
            roomId: roomId,
            fromPeerId: fromPeerId
        )
    }
    
    // MARK: - WebRTC Operations
    
    public func createPeerConnection() async throws -> RTCPeerConnection {
        return try await webRTCManager.createPeerConnection()
    }
    
    public func createDataChannel(_ connection: RTCPeerConnection, label: String = "data") -> RTCDataChannel? {
        let config = RTCDataChannelConfiguration()
        config.isOrdered = true
        return connection.dataChannel(forLabel: label, configuration: config)
    }
    
    public func createOffer(for connection: RTCPeerConnection) async throws -> RTCSessionDescription {
        return try await webRTCManager.createOffer(for: connection)
    }
    
    public func createAnswer(for connection: RTCPeerConnection) async throws -> RTCSessionDescription {
        return try await webRTCManager.createAnswer(for: connection)
    }
    
    public func setLocalDescription(_ sdp: RTCSessionDescription, for connection: RTCPeerConnection) async throws {
        try await webRTCManager.setLocalDescription(sdp, for: connection)
    }
    
    public func setRemoteDescription(_ sdp: RTCSessionDescription, for connection: RTCPeerConnection) async throws {
        try await webRTCManager.setRemoteDescription(sdp, for: connection)
    }
    
    public func addIceCandidate(_ candidate: RTCIceCandidate, to connection: RTCPeerConnection) async throws {
        try await webRTCManager.addIceCandidate(candidate, to: connection)
    }
}

// MARK: - iOS Specific Extensions

#if os(iOS)
import UIKit

public extension P2P2Core {
    /// Configure for iOS app usage
    func configureForIOS() {
        // Configure audio session for WebRTC
        let audioSession = AVAudioSession.sharedInstance()
        do {
            try audioSession.setCategory(.playAndRecord, mode: .voiceChat, options: [.allowBluetooth, .allowBluetoothA2DP])
            try audioSession.setActive(true)
        } catch {
            print("Failed to configure audio session: \(error)")
        }
    }
    
    /// Handle app lifecycle for iOS
    func handleAppDidEnterBackground() {
        // Reduce discovery interval or pause
    }
    
    func handleAppWillEnterForeground() {
        // Resume normal discovery
    }
}
#endif

// MARK: - macOS Specific Extensions

#if os(macOS)
import AppKit

public extension P2P2Core {
    /// Configure for macOS app usage
    func configureForMacOS() {
        // macOS specific configuration if needed
    }
}
#endif